### PR TITLE
grant/kol 7429 programmatic retrieval of significance

### PR DESCRIPTION
**Pull request by grant-Kolena**
- add functionality to display performance delta when using download_quality_standard_result
- added test

### Please check if the PR fulfills these requirements

- [ ] Include reference to internal ticket and/or GitHub issue "Fixes #NNNN" (if applicable)
- [ ] Relevant tests for the changes have been added
- [ ] Relevant docs have been added / updated

### CODE REVIEW

File: kolena/_api/v2/_api.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
+# Copyright 2021-2024 Kolena Inc.
```
Comment (grant-Kolena) [inline kolena/_api/v2/_api.py:1]: Api models copied from Kolena Backend code base to support necessary API calls

File: kolena/_experimental/utils.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    Adapted from a polynomial approximation in:
+            Ibbetson D, Algorithm 209
+            Collected Algorithms of the CACM 1963 p. 616
+    Note:
+            This routine has six digit accuracy, so it is only useful for absolute
+            z values <= 6.  For z values > to 6.0, poz() returns 0.0.
+"""
+
+
+def z_score_to_probability(z: float) -> float:
```
Comment (grant-Kolena) [inline kolena/_experimental/utils.py:36]: Copied and rewritten from kolena FE's implementation of the same functionality

File: kolena/_experimental/quality_standard.py [hunk 3]
```diff
@@ -11,26 +11,57 @@
...
 from kolena._utils import log
 from kolena._utils.instrumentation import with_event
+from kolena._utils.serde import from_dict
 from kolena.dataset.dataset import _load_dataset_metadata
+from kolena.dataset.evaluation import _get_eval_config_id
+from kolena.dataset.evaluation import _get_model_id
 from kolena.errors import IncorrectUsageError
 
 
+class PerformanceDelta(str, enum.Enum):
```
Comment (munkyshi) [inline kolena/_experimental/quality_standard.py]: Do these serialize well to CSV? Will users ever be expected to know about `PerformanceDelta`? In general, for anything that a user deals with, I'd prefer to use something like
```
PerformanceDelta = Literal["improved", "regressed", "similar", "unknown"]
```

So that the user doesn't need to know about some random enum type, and can instead just work with things as strings.

Comment (grant-Kolena) [reply kolena/_experimental/quality_standard.py]: I can confirm export to CSV works fine, but will still change to Literal so the user don't have to see the enum

File: kolena/_experimental/quality_standard.py [hunk 4]
```diff
@@ -87,17 +295,38 @@ def download_quality_standard_result(
     :param metric_groups: The names of the metric groups to include in the result.
     :param intersect_results: If True, only include datapoint that are common to all models in the metrics calculation.
     Note all metric groups are included when this value is `None`.
+    :param confidence_level: The confidence score used to calculate the Margin of Error, representing the probability of
+     capturing the true population parameter within the calculated MOE, we recommend setting between 0.9 - 0.99.
+     If this is specified performance delta column will be shown, metric valued will be classified as
+      improved, regressed, similar or unknown according to the MOE.
+    :param reference_model: The name of the model to use as a reference for the Margin of Error calculation,
```
Comment (munkyshi) [inline kolena/_experimental/quality_standard.py]:
```suggestion
    :param reference_model: The name of the model to use as a reference for the Margin of Error calculation.
        This should be one of the models provided in `models`. If unspecified, the first model of `models` will be used as reference.
```

File: kolena/_experimental/quality_standard.py [hunk 5]
```diff
@@ -87,17 +295,38 @@ def download_quality_standard_result(
...
+    :param reference_eval_config: The evaluation configuration to use in conjunction with the reference model,
+     if unspecified the first model in the models list will be used
+     as a reference for the Margin of Error, for None eval config, pass in 'null' instead of None. If this is None,
+     the first eval config will be used.
```
Comment (munkyshi) [inline kolena/_experimental/quality_standard.py]:
```suggestion
    :param reference_eval_config: The evaluation configuration to use in conjunction with the `reference_model`.
        If this is `None`, the first eval config will be used. If you wish to use an evaluation configuration of `None`, please instead specify the configuration as `'null'`.
```

Also, seems a bit odd to make the user specify `null` instead of using `None` -- would it make sense to instead define it as:
```
reference_eval_config: Optional[Union[Dict[str, Any], Literal["first"]]] = "first",
```

File: kolena/_api/v2/_metric.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+        return v or "macro"
+
+
+@dataclass(frozen=True, order=True)
+class RegressionAggregator:
+    aggregator: Literal["mae", "mse", "rmse", "r_squared", "pearson_corr", "spearman_corr"]
+
+
+@dataclass(frozen=True, order=True)
+class ClassificationParams:
```
Comment (munkyshi) [inline kolena/_api/v2/_metric.py:94]: How did we end up with a mix of camelCase and snake_case? It's sloppy to have this in our API.

Comment (munkyshi) [reply kolena/_api/v2/_metric.py:94]: Not sure if there's anything to do about this -- do we need all of these classes though? Really annoying to make this public.

File: tests/integration/_experimental/test_quality_standard.py [hunk 7]
```diff
@@ -282,6 +282,169 @@ def test__download_quality_standard_result__union(
     )
 
 
+def test__download_quality_standard_result_with_performance_delta(
+    datapoints: pd.DataFrame,
+    results: List[Tuple[EvalConfig, pd.DataFrame]],
+) -> None:
+    dataset_name = with_test_prefix("test__download_quality_standard_result_with_performance_delta")
+    model_name = with_test_prefix("test__test__download_quality_standard_result_with_performance_delta_model")
```
Comment (munkyshi) [inline tests/integration/_experimental/test_quality_standard.py]:
```suggestion
    model_name = with_test_prefix("test__download_quality_standard_result_with_performance_delta_model")
```

**Comment by munkyshi** [comment]
Can you share an example of how to run this? I'm trying to figure out how to best set up a dataset / QS for this -- is there an example that you already have set up that I can play with?

**Comment by grant-Kolena** [comment]
> Can you share an example of how to run this? I'm trying to figure out how to best set up a dataset / QS for this -- is there an example that you already have set up that I can play with?

Yes, I have played around with this dataset on my tenant

```
from kolena._experimental import download_quality_standard_result
qs_result = download_quality_standard_result("attack_on_llms_small", ["databricks_dbrx-instruct", "gpt-4-turbo"], confidence_level = 0.95, reference_model = "gpt-4-turbo")
```

File: kolena/_experimental/quality_standard.py [hunk 8]
```diff
@@ -87,17 +295,38 @@ def download_quality_standard_result(
...
     :return: A Dataframe containing the quality standard result.
     """
+    if reference_model and reference_model not in models:
+        raise IncorrectUsageError("reference model not in models")
```
Comment (munkyshi) [inline kolena/_experimental/quality_standard.py]: We can improve the message here:
```suggestion
        raise IncorrectUsageError(f"The specified reference model '{reference_model}' was not one of the provided models")
```

File: kolena/_experimental/quality_standard.py [hunk 9]
```diff
@@ -72,12 +97,195 @@ def _download_quality_standard_result(
...
+    qs_result: pd.DataFrame,
+    moe: Dict[Tuple[str, Any], float],
+    reference_model: Optional[str] = None,
+    reference_eval_config: Union[Dict[str, Any], Literal["null"], None] = None,
+) -> pd.DataFrame:
+    for model, eval_config, _, _ in qs_result.columns:
+        if not reference_model:
+            reference_model = model
+        if reference_model == model and not reference_eval_config:
+            reference_eval_config = eval_config
```
Comment (munkyshi) [inline kolena/_experimental/quality_standard.py]: Does this mean that the first `eval_config` will be skipped if it is `None`?

```
# initially reference_model is None and reference_eval_config
# First pass, model is 'foo' and eval_config is None
reference_model = 'foo'
reference_eval_config = None
# Second pass, model is 'foo' and eval_config is {'bar': 'baz'}
reference_eval_config = {'bar': 'baz'}
```

Also, does this get handled safely if the only eval config for the reference model is `None`?

File: kolena/_experimental/quality_standard.py (hunk 5 above)
Comment (grant-Kolena) [reply kolena/_experimental/quality_standard.py]: Yep make sense, I was just trying to distinguish between null and None

File: kolena/_api/v2/_metric.py (hunk 6 above)
Comment (grant-Kolena) [reply kolena/_api/v2/_metric.py:94]: I just copied everything from the server side, We need the QS and Testing endpoint to function properly which involves all the complex Filter models. We could simplify this to strictly what we need and and the additional stuff later, but I don't think this is removing too much of the models as there can be filters in stratifications of a qs.

Comment (munkyshi) [reply kolena/_api/v2/_metric.py:94]: Yeah, this is frustrating, but technically still non-public under `_api` -- just frustrating, but not worth changing.

File: kolena/_experimental/quality_standard.py (hunk 9 above)
Comment (grant-Kolena) [reply kolena/_experimental/quality_standard.py]: originally I was expecting user to pass in "null" so it won't be ignored, I will have to handle this case now. This should work as expected when the only eval config is None, as the dataset I used to test is like this

File: kolena/_api/v2/_api.py (hunk 1 above)
Comment (munkyshi) [reply kolena/_api/v2/_api.py:1]: Will we need to add these to https://github.com/kolenaIO/shipyard/tree/trunk/sdk-server-api/sdk_server_api/v2

Comment (grant-Kolena) [reply kolena/_api/v2/_api.py:1]: This setup is kinda messy, FE hit galleon for QS, galleon QS endpoint is open to sdk server for sdk client access through gateway. So we dont actually need them in sdk-server

**Comment by munkyshi** [comment]
> ```
> from kolena._experimental import download_quality_standard_result
> qs_result = download_quality_standard_result("attack_on_llms_small", ["databricks_dbrx-instruct", "gpt-4-turbo"], confidence_level = 0.95, reference_model = "gpt-4-turbo")
> ```

I tried playing around with this, and it looks like the colors aren't matching with the webapp ([link](https://trunk.kolena.io/grant-tenant/dataset/standards?datasetId=228&models=N4IglgJiBcCMDM8A0IDGB7AdgMzAcwFcAnAQwBcwsZRIZYBfFTEgWwFMYQ8AHMgWgAsfMsQBG6EPSA&models=N4IglgJiBcCMDMAmANCAxgewHYDMwHMBXAJwEMAXMbGUSGWAXwaA&metricGroupVisibilities=N4IgbglgzhBGA2BTEAuALgJwK6IDQgFtFMIBjKVAbVEhgWXW0QF9cbo4lVMdmBdZkA))

One example (from when I dumped this to CSV):
```
model,,databricks_dbrx-instruct,databricks_dbrx-instruct,gpt-4-turbo,gpt-4-turbo,databricks_dbrx-instruct,databricks_dbrx-instruct,gpt-4-turbo,gpt-4-turbo
eval_config,,null,null,null,null,null,null,null,null
metric_group,,Metric Group,Metric Group,Metric Group,Metric Group,Metric Group,Metric Group,Metric Group,Metric Group
metric,,rate rejected,Mean Metadata.count,rate rejected,Mean Metadata.count,rate rejected,Mean Metadata.count,rate rejected,Mean Metadata.count
type,,metric_value,metric_value,metric_value,metric_value,performance_delta,performance_delta,performance_delta,performance_delta
stratification,test_case,,,,,,,,
...
type,Raw,0.03333333333333333,,0.1,0.0,regressed,unknown,similar,similar
type,Role Playing,0.15555555555555556,0.6666666666666666,0.12222222222222222,,improved,unknown,similar,unknown
```
vs webapp
Screenshot 2024-10-09 at 17 34 46

The endpoint seems to indicate regressed and improved, but there's no coloration on the webapp (both at confidence level of 0.95)

**Comment by grant-Kolena** [comment]
> > ```
> > from kolena._experimental import download_quality_standard_result
> > qs_result = download_quality_standard_result("attack_on_llms_small", ["databricks_dbrx-instruct", "gpt-4-turbo"], confidence_level = 0.95, reference_model = "gpt-4-turbo")
> > ```
> 
> I tried playing around with this, and it looks like the colors aren't matching with the webapp ([link](https://trunk.kolena.io/grant-tenant/dataset/standards?datasetId=228&models=N4IglgJiBcCMDM8A0IDGB7AdgMzAcwFcAnAQwBcwsZRIZYBfFTEgWwFMYQ8AHMgWgAsfMsQBG6EPSA&models=N4IglgJiBcCMDMAmANCAxgewHYDMwHMBXAJwEMAXMbGUSGWAXwaA&metricGroupVisibilities=N4IgbglgzhBGA2BTEAuALgJwK6IDQgFtFMIBjKVAbVEhgWXW0QF9cbo4lVMdmBdZkA))
> 
> One example (from when I dumped this to CSV):
> 
> ```
> model,,databricks_dbrx-instruct,databricks_dbrx-instruct,gpt-4-turbo,gpt-4-turbo,databricks_dbrx-instruct,databricks_dbrx-instruct,gpt-4-turbo,gpt-4-turbo
> eval_config,,null,null,null,null,null,null,null,null
> metric_group,,Metric Group,Metric Group,Metric Group,Metric Group,Metric Group,Metric Group,Metric Group,Metric Group
> metric,,rate rejected,Mean Metadata.count,rate rejected,Mean Metadata.count,rate rejected,Mean Metadata.count,rate rejected,Mean Metadata.count
> type,,metric_value,metric_value,metric_value,metric_value,performance_delta,performance_delta,performance_delta,performance_delta
> stratification,test_case,,,,,,,,
> ...
> type,Raw,0.03333333333333333,,0.1,0.0,regressed,unknown,similar,similar
> type,Role Playing,0.15555555555555556,0.6666666666666666,0.12222222222222222,,improved,unknown,similar,unknown
> ```
> 
> vs webapp Screenshot 2024-10-09 at 17 34 46
> 
> The endpoint seems to indicate regressed and improved, but there's no coloration on the webapp (both at confidence level of 0.95)

Weird, the algorithm is exactly the same, I will calculate it again to confirm. Could it be that the color on webapp is too light to see

**Comment by grant-Kolena** [comment]
> > > ```
> > > from kolena._experimental import download_quality_standard_result
> > > qs_result = download_quality_standard_result("attack_on_llms_small", ["databricks_dbrx-instruct", "gpt-4-turbo"], confidence_level = 0.95, reference_model = "gpt-4-turbo")
> > > ```
> > 
> > 
> > I tried playing around with this, and it looks like the colors aren't matching with the webapp ([link](https://trunk.kolena.io/grant-tenant/dataset/standards?datasetId=228&models=N4IglgJiBcCMDM8A0IDGB7AdgMzAcwFcAnAQwBcwsZRIZYBfFTEgWwFMYQ8AHMgWgAsfMsQBG6EPSA&models=N4IglgJiBcCMDMAmANCAxgewHYDMwHMBXAJwEMAXMbGUSGWAXwaA&metricGroupVisibilities=N4IgbglgzhBGA2BTEAuALgJwK6IDQgFtFMIBjKVAbVEhgWXW0QF9cbo4lVMdmBdZkA))
> > One example (from when I dumped this to CSV):
> > ```
> > model,,databricks_dbrx-instruct,databricks_dbrx-instruct,gpt-4-turbo,gpt-4-turbo,databricks_dbrx-instruct,databricks_dbrx-instruct,gpt-4-turbo,gpt-4-turbo
> > eval_config,,null,null,null,null,null,null,null,null
> > metric_group,,Metric Group,Metric Group,Metric Group,Metric Group,Metric Group,Metric Group,Metric Group,Metric Group
> > metric,,rate rejected,Mean Metadata.count,rate rejected,Mean Metadata.count,rate rejected,Mean Metadata.count,rate rejected,Mean Metadata.count
> > type,,metric_value,metric_value,metric_value,metric_value,performance_delta,performance_delta,performance_delta,performance_delta
> > stratification,test_case,,,,,,,,
> > ...
> > type,Raw,0.03333333333333333,,0.1,0.0,regressed,unknown,similar,similar
> > type,Role Playing,0.15555555555555556,0.6666666666666666,0.12222222222222222,,improved,unknown,similar,unknown
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > vs webapp Screenshot 2024-10-09 at 17 34 46
> > The endpoint seems to indicate regressed and improved, but there's no coloration on the webapp (both at confidence level of 0.95)
> 
> Weird, the algorithm is exactly the same, I will calculate it again to confirm. Could it be that the color on webapp is too light to see

this matches with the UI now, not sure what type of test would be meaningful to these mathematical calculations, it is hard to identify edge cases here. We already have a test case to check if the produced performance delta match the expected result.

**Comment by munkyshi** [comment]
> this matches with the UI now, not sure what type of test would be meaningful to these mathematical calculations, it is hard to identify edge cases here. We already have a test case to check if the produced performance delta match the expected result.

What was the issue, the check for `and max_value > 0:`? Is there anything we can do to test around that case specifically, since it was something that was missed? Change looks good otherwise.

**Review by munkyshi — APPROVED** [review]
Approving to unblock, pending test to track the latest (now fixed) issue.

**Comment by grant-Kolena** [comment]
> > this matches with the UI now, not sure what type of test would be meaningful to these mathematical calculations, it is hard to identify edge cases here. We already have a test case to check if the produced performance delta match the expected result.
> 
> What was the issue, the check for `and max_value > 0:`? Is there anything we can do to test around that case specifically, since it was something that was missed? Change looks good otherwise.

I missed the part where the max value for performance delta is the max across the stratification, except for when the max value is 0 or less.

**Comment by grant-Kolena** [comment]
> Approving to unblock, pending test to track the latest (now fixed) issue.

created this [ticket](https://linear.app/kolena/issue/KOL-7626/meaningful-test-for-client-performance-delta) to track
